crates/adapters/src/controller/mod.rs

```diff
@@ -472,321 +457,946 @@ impl Controller {
+            || buffered_records > self.min_batch_size_records
+            || tick.map_or(false, |t| Instant::now() >= t)
+            || self.buffer_timeout.map_or(false, |t| Instant::now() >= t)
+        {
+            self.tick = self.clock_resolution.map(|delay| Instant::now() + delay);
+            self.buffer_timeout = None;
+            None
+        } else {
+            if buffered_records > 0 && self.buffer_timeout.is_none() {
+                self.buffer_timeout = Some(Instant::now() + self.max_buffering_delay);
+            }
+            match (tick, self.buffer_timeout) {
+                (Some(t1), Some(t2)) => Some(Wait::Time(min(t1, t2))),
+                (Some(t), None) | (None, Some(t)) => Some(Wait::Time(t)),
+                (None, None) => Some(Wait::Forever),
+            }
+        }
+    }
+}
+
+/// Controller initialization.
+///
+/// When we start a controller, we do one of:
+///
+/// - Start from an existing checkpoint.
+///
+/// - Start a new pipeline while a new initial checkpoint.
+///
+/// - Start a new pipeline without any checkpoint support.
+///
+/// This structure handles all these cases.
+struct ControllerInit {
+    /// The pipeline configuration.
+    ///
+    /// This will differ from the one passed into [ControllerInit::new] if a
+    /// checkpoint is read, because a checkpoint includes the pipeline
+    /// configuration.
+    pipeline_config: PipelineConfig,
+
+    /// The circuit configuration.
+    circuit_config: CircuitConfig,
+
+    /// Fault-tolerance initialization, if FT will be enabled.
+    ft: Option<FtInit>,
+}
+
+struct FtInit {
+    /// The first step that the circuit will execute.
+    step: Step,
+
+    /// The path to the `state.json` checkpoint file, if fault tolerance is
+    /// enabled.
+    state_path: PathBuf,
+
+    /// The step reader/writer, if fault tolerance is enabled.
+    step_rw: StepRw,
+}
+
+impl ControllerInit {
+    fn new(config: PipelineConfig) -> Result<Self, ControllerError> {
+        let Some(ft) = config.global.fault_tolerance else {
+            info!("fault tolerance is disabled in configuration");
+            return Self::without_ft(config);
+        };
+        let Some(path) = config.storage_config.as_ref().map(|storage| storage.path()) else {
+            return Err(ControllerError::Config {
+                config_error: ConfigError::FtRequiresStorage,
+            });
+        };
+
+        fn startup_io_error(error: IoError) -> ControllerError {
+            ControllerError::io_error(String::from("controller startup"), error)
+        }
+        let state_path = path.join("state.json");
+        let steps_path = path.join("steps.bin");
+
+        // If we're allowed to resume from a checkpoint and one exists, use it.
+        if ft == FtConfig::LatestCheckpoint && fs::exists(&state_path).map_err(startup_io_error)? {
+            // Open the existing checkpoint.
+            info!(
+                "{}: initializing fault tolerance from saved state",
```

Comment:
Would a better message something along the lines of "initializing pipeline from a saved checkpoint"?